fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Compensation.java

```diff
@@ -709,7 +709,7 @@ public WithSelectCompensation derivedWithPredicateCompensationMap(final boolean
                                                                           @Nonnull final Set<? extends Quantifier> unmatchedQuantifiers,
                                                                           @Nonnull final Set<CorrelationIdentifier> compensatedAliases,
                                                                           @Nonnull final Optional<Value> remainingComputationValueOptional) {
-            Verify.verify(!predicateCompensationMap.isEmpty() || remainingComputationValueOptional.isPresent());
```

Comment:
For this particular case, the creation of the `Compensation` in `SelectExpression#compensate` is relevant for signalling the need of `Exists` to inject itself as a compensating `QueryPredicate` in the containing query fragment as in here: 

https://github.com/FoundationDB/fdb-record-layer/blob/c504a045f318da8bef5c611d9fa62579610c9a4f/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/predicates/ExistsPredicate.java#L212-L215

Moreover, a `Compensation` is something that is supposed to filter more rows and/or change the shape of the results, I think creating a compensation without a predicate nor a remaining computation is malformed, for more information, looks at the `apply` logic of `ForMatch`:

https://github.com/FoundationDB/fdb-record-layer/blob/c504a045f318da8bef5c611d9fa62579610c9a4f/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Compensation.java#L726-L735

we return immediately if we do not find a predicate in the compensation predicate map.

and therefore, if in your case, when we notice an unmatched quantifier only, we should probably create an impossible compensation instead, which, in addition to what I mentioned above, more efficient, considering that it is only used for triggering the injection of `Exists` itself as a compensation.

Reply:
My suggestion is to change this block:

https://github.com/FoundationDB/fdb-record-layer/blob/c504a045f318da8bef5c611d9fa62579610c9a4f/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/expressions/SelectExpression.java#L746-L751

to become something like this:

```java
final var isCompensationNeeded =
        !predicateCompensationMap.isEmpty() || matchInfo.getRemainingComputationValueOptional().isPresent();

if (!isCompensationNeeded) {
    if (!unmatchedQuantifiers.isEmpty()) {
        return Compensation.impossibleCompensation();
    }
    return Compensation.noCompensation();
}
```